{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c126d8d3-8f13-423b-9599-8cdbcc78fce6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "works\n"
     ]
    }
   ],
   "source": [
    "# import relavent packages\n",
    "import pandas as pd\n",
    "import requests\n",
    "import ftplib\n",
    "import os\n",
    "import ftplib\n",
    "import numpy as np\n",
    "print('works') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a1c77b0b-044a-438b-bec8-851ccae9a8fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connected to FTP server: ftp.ebi.ac.uk\n",
      "Changed to directory: pub/databases/gwas/summary_statistics/GCST003001-GCST004000/GCST003659/\n",
      "Files in directory pub/databases/gwas/summary_statistics/GCST003001-GCST004000/GCST003659/: ['MAGIC ISI README.docx', 'MAGIC_ISI_Model_3_JMA.txt', 'harmonised']\n",
      "File MAGIC_ISI_Model_3_JMA.txt exists in the directory.\n",
      "Connected to FTP server: ftp.ebi.ac.uk\n",
      "Changed to directory: pub/databases/gwas/summary_statistics/GCST003001-GCST004000/GCST003659/\n",
      "File GCST003659_MAGIC_ISI_Model_3_JMA.txt downloaded successfully.\n",
      "Downloaded file GCST003659_MAGIC_ISI_Model_3_JMA.txt exists.\n"
     ]
    }
   ],
   "source": [
    "# looks for the file online\n",
    "ftp_server = 'ftp.ebi.ac.uk'\n",
    "ftp_base_path = 'pub/databases/gwas/summary_statistics/GCST003001-GCST004000/GCST003659/'\n",
    "accession_id = 'GCST003659'\n",
    "ftp_file_name = 'MAGIC_ISI_Model_3_JMA.txt'\n",
    "\n",
    "try:\n",
    "    with ftplib.FTP(ftp_server) as ftp:\n",
    "        ftp.login()\n",
    "        print(f\"Connected to FTP server: {ftp_server}\")\n",
    "        ftp.cwd(ftp_base_path)\n",
    "        print(f\"Changed to directory: {ftp_base_path}\")\n",
    "        files = ftp.nlst()\n",
    "        print(f\"Files in directory {ftp_base_path}: {files}\")\n",
    "        if ftp_file_name in files:\n",
    "            print(f\"File {ftp_file_name} exists in the directory.\")\n",
    "        else:\n",
    "            print(f\"File {ftp_file_name} does not exist in the directory.\")\n",
    "except ftplib.all_errors as e:\n",
    "    print(f\"FTP error: {e}\")\n",
    "\n",
    "# now to actually download the file...\n",
    "local_filename = f'{accession_id}_{ftp_file_name}'\n",
    "\n",
    "try:\n",
    "    with ftplib.FTP(ftp_server) as ftp:\n",
    "        ftp.login()\n",
    "        print(f\"Connected to FTP server: {ftp_server}\")\n",
    "        ftp.cwd(ftp_base_path)\n",
    "        print(f\"Changed to directory: {ftp_base_path}\")\n",
    "        with open(local_filename, 'wb') as local_file:\n",
    "            ftp.retrbinary(f'RETR {ftp_file_name}', local_file.write)\n",
    "    print(f\"File {local_filename} downloaded successfully.\")\n",
    "except ftplib.all_errors as e:\n",
    "    print(f\"FTP error: {e}\")\n",
    "\n",
    "if os.path.exists(local_filename):\n",
    "    print(f\"Downloaded file {local_filename} exists.\")\n",
    "else:\n",
    "    print(f\"Downloaded file {local_filename} does not exist.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "7f1dd56e-460b-49fb-825c-ab81e202cec5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File loaded successfully.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr_build36</th>\n",
       "      <th>pos_build36</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>maf_hapmap_rel27</th>\n",
       "      <th>effect</th>\n",
       "      <th>stderr</th>\n",
       "      <th>pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>742429</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.165</td>\n",
       "      <td>-0.097</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.4250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs3131968</td>\n",
       "      <td>1</td>\n",
       "      <td>744055</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.130</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.2233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs12124819</td>\n",
       "      <td>1</td>\n",
       "      <td>766409</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.4208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs2980319</td>\n",
       "      <td>1</td>\n",
       "      <td>766985</td>\n",
       "      <td>a</td>\n",
       "      <td>t</td>\n",
       "      <td>0.134</td>\n",
       "      <td>0.092</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.2842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs2905062</td>\n",
       "      <td>1</td>\n",
       "      <td>774913</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.144</td>\n",
       "      <td>-0.077</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.3403</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          snp  chr_build36  pos_build36 effect_allele other_allele  \\\n",
       "0   rs3094315            1       742429             a            g   \n",
       "1   rs3131968            1       744055             a            g   \n",
       "2  rs12124819            1       766409             a            g   \n",
       "3   rs2980319            1       766985             a            t   \n",
       "4   rs2905062            1       774913             a            g   \n",
       "\n",
       "   maf_hapmap_rel27  effect  stderr  pvalue  \n",
       "0             0.165  -0.097    0.11  0.4250  \n",
       "1             0.125   0.130    0.12  0.2233  \n",
       "2             0.222   0.081    0.12  0.4208  \n",
       "3             0.134   0.092    0.13  0.2842  \n",
       "4             0.144  -0.077    0.13  0.3403  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['snp', 'chr_build36', 'pos_build36', 'effect_allele', 'other_allele',\n",
      "       'maf_hapmap_rel27', 'effect', 'stderr', 'pvalue'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# check if the file was actually downloaded, this is mostly for debugging\n",
    "if os.path.exists(local_filename):\n",
    "    try:\n",
    "        # load the downloaded gwas summary stats file\n",
    "        gwas_summary = pd.read_csv(local_filename, delimiter='\\t')\n",
    "        print(\"File loaded successfully.\")\n",
    "        # display the first few rows and column names for debugging purposes\n",
    "        display(gwas_summary.head())\n",
    "        print(gwas_summary.columns)\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred while loading the file: {e}\")\n",
    "else:\n",
    "    print(f\"File {local_filename} not found.\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "be05cd43-b3d2-4452-bb2c-a392fe0e829c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GWAS Summary Statistics:\n",
      "          snp  chr_build36  pos_build36 effect_allele other_allele  \\\n",
      "0   rs7517384            1     80017509             t            c   \n",
      "1   rs1857095            1     80018608             t            c   \n",
      "2  rs10495667            2     17686973             a            g   \n",
      "3   rs7578326            2    226728897             a            g   \n",
      "4  rs13405357            2    226730280             t            c   \n",
      "\n",
      "   maf_hapmap_rel27  effect  stderr        pvalue  \n",
      "0             0.033  -0.880   0.250  9.605000e-16  \n",
      "1             0.034   0.880   0.250  5.398000e-16  \n",
      "2             0.040  -0.900   0.180  3.826000e-09  \n",
      "3             0.350   0.076   0.079  1.210000e-08  \n",
      "4             0.350   0.100   0.090  2.145000e-08  \n",
      "Number of common SNPs: 10\n",
      "Filtered GWAS Summary (first 5 rows):\n",
      "          snp  chr_build36  pos_build36 effect_allele other_allele  \\\n",
      "0   rs7517384            1     80017509             t            c   \n",
      "1   rs1857095            1     80018608             t            c   \n",
      "2  rs10495667            2     17686973             a            g   \n",
      "3   rs7578326            2    226728897             a            g   \n",
      "4  rs13405357            2    226730280             t            c   \n",
      "\n",
      "   maf_hapmap_rel27  effect  stderr        pvalue  \n",
      "0             0.033  -0.880   0.250  9.605000e-16  \n",
      "1             0.034   0.880   0.250  5.398000e-16  \n",
      "2             0.040  -0.900   0.180  3.826000e-09  \n",
      "3             0.350   0.076   0.079  1.210000e-08  \n",
      "4             0.350   0.100   0.090  2.145000e-08  \n",
      "Filtered Genotype Data (first 5 rows):\n",
      "         rsID  chromosome   position effect_allele other_allele  \\\n",
      "0   rs7517384           1   80017509             t            c   \n",
      "1   rs1857095           1   80018608             t            c   \n",
      "2  rs10495667           2   17686973             a            g   \n",
      "3   rs7578326           2  226728897             a            g   \n",
      "4  rs13405357           2  226730280             t            c   \n",
      "\n",
      "   maf_hapmap_rel27  effect  stderr        pvalue genotype  \n",
      "0             0.033  -0.880   0.250  9.605000e-16       AA  \n",
      "1             0.034   0.880   0.250  5.398000e-16       AA  \n",
      "2             0.040  -0.900   0.180  3.826000e-09       AG  \n",
      "3             0.350   0.076   0.079  1.210000e-08       GG  \n",
      "4             0.350   0.100   0.090  2.145000e-08       TT  \n"
     ]
    }
   ],
   "source": [
    "# load gwas summary stats and genotype data from poscontrol.sh file\n",
    "gwas_summary = pd.read_csv('poscontrol.sh', delim_whitespace=True)\n",
    "\n",
    "# shows gwas stats for debugging purposes\n",
    "print(\"GWAS Summary Statistics:\")\n",
    "print(gwas_summary.head())\n",
    "\n",
    "# normalize snps in gwas summary stats\n",
    "gwas_summary['snp'] = gwas_summary['snp'].str.strip().str.lower()\n",
    "\n",
    "# making the dataframe\n",
    "genotype_df = gwas_summary[['snp', 'chr_build36', 'pos_build36', 'effect_allele', 'other_allele', 'maf_hapmap_rel27', 'effect', 'stderr', 'pvalue']]\n",
    "\n",
    "genotype_df.columns = ['rsID', 'chromosome', 'position', 'effect_allele', 'other_allele', 'maf_hapmap_rel27', 'effect', 'stderr', 'pvalue']\n",
    "\n",
    "genotype_df['genotype'] = ['AA', 'AA', 'AG', 'GG', 'TT', 'TT', 'TC', 'CT', 'AA', 'GG', 'AA', 'AA', 'AG', 'GG', 'TT', 'TT', 'TC', 'CT', 'AA', 'GG']\n",
    "\n",
    "# normalize snp ids in the genotype data\n",
    "genotype_df['rsID'] = genotype_df['rsID'].str.strip().str.lower()\n",
    "\n",
    "# find common snps from signifcant snps from genotype data\n",
    "common_snps = set(gwas_summary['snp']).intersection(set(genotype_df['rsID']))\n",
    "print(f\"Number of common SNPs: {len(common_snps)}\")\n",
    "\n",
    "# filter gwas summary statistics and genotype data to include only common snps\n",
    "filtered_gwas_summary = gwas_summary[gwas_summary['snp'].isin(common_snps)]\n",
    "filtered_genotype_df = genotype_df[genotype_df['rsID'].isin(common_snps)]\n",
    "\n",
    "print(f\"Filtered GWAS Summary (first 5 rows):\\n{filtered_gwas_summary.head()}\")\n",
    "print(f\"Filtered Genotype Data (first 5 rows):\\n{filtered_genotype_df.head()}\")\n",
    "\n",
    "# function to convert genotype to numeric value\n",
    "def genotype_to_numeric(genotype):\n",
    "    genotype_map = {'AA': 0, 'AG': 1, 'GA': 1, 'GG': 2, 'TT': 0, 'TC': 1, 'CT': 1, 'CC': 2}\n",
    "    return genotype_map.get(genotype, np.nan)\n",
    "\n",
    "# apply the conversion function to the genotype data\n",
    "filtered_genotype_df['genotype_numeric'] = filtered_genotype_df['genotype'].apply(genotype_to_numeric)\n",
    "\n",
    "# convert filtered genotype data to dictionary for prs calc\n",
    "genotype_data_dict = pd.Series(filtered_genotype_df['genotype_numeric'].values, index=filtered_genotype_df['rsID']).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "10421491-9ab9-4397-a326-db4546d845a9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of significant SNPs: 20\n",
      "SNP ID: rs7517384, Effect Size: -0.88, Genotype: 0\n",
      "SNP ID: rs1857095, Effect Size: 0.88, Genotype: 0\n",
      "SNP ID: rs10495667, Effect Size: -0.9, Genotype: 1\n",
      "SNP ID: rs7578326, Effect Size: 0.076, Genotype: 2\n",
      "SNP ID: rs13405357, Effect Size: 0.1, Genotype: 0\n",
      "SNP ID: rs6092841, Effect Size: -0.53, Genotype: 0\n",
      "SNP ID: rs2828532, Effect Size: 1.1, Genotype: 1\n",
      "SNP ID: rs2828535, Effect Size: 1.1, Genotype: 1\n",
      "SNP ID: rs2828537, Effect Size: 1.1, Genotype: 0\n",
      "SNP ID: rs10483182, Effect Size: -2.0, Genotype: 2\n",
      "SNP ID: rs7517384, Effect Size: -0.88, Genotype: 0\n",
      "SNP ID: rs1857095, Effect Size: 0.88, Genotype: 0\n",
      "SNP ID: rs10495667, Effect Size: -0.9, Genotype: 1\n",
      "SNP ID: rs7578326, Effect Size: 0.076, Genotype: 2\n",
      "SNP ID: rs13405357, Effect Size: 0.1, Genotype: 0\n",
      "SNP ID: rs6092841, Effect Size: -0.53, Genotype: 0\n",
      "SNP ID: rs2828532, Effect Size: 1.1, Genotype: 1\n",
      "SNP ID: rs2828535, Effect Size: 1.1, Genotype: 1\n",
      "SNP ID: rs2828537, Effect Size: 1.1, Genotype: 0\n",
      "SNP ID: rs10483182, Effect Size: -2.0, Genotype: 2\n",
      "Polygenic Risk Score: -5.096\n"
     ]
    }
   ],
   "source": [
    "# calculates the PRS \n",
    "def calculate_prs(gwas_summary, genotype_data_dict, p_value_col, snp_col='snp', effect_size_col='effect'):\n",
    "    # filter for significant snps (p-value < 5e-8)\n",
    "    significant_snps = gwas_summary[gwas_summary[p_value_col] < 5e-8]\n",
    "    print(f\"Number of significant SNPs: {len(significant_snps)}\")\n",
    "    \n",
    "    # initialize prs score\n",
    "    prs_score = 0\n",
    "    \n",
    "    # loop thru significant snps and calc the prs\n",
    "    for index, snp in significant_snps.iterrows():\n",
    "        snp_id = snp[snp_col]\n",
    "        effect_size = snp[effect_size_col]\n",
    "        \n",
    "        # see if snp is in genotype data\n",
    "        if snp_id in genotype_data_dict:\n",
    "            genotype = genotype_data_dict[snp_id]\n",
    "        else:\n",
    "            print(f\"SNP ID {snp_id} not found in genotype data. Skipping.\")\n",
    "            continue\n",
    "        \n",
    "        print(f\"SNP ID: {snp_id}, Effect Size: {effect_size}, Genotype: {genotype}\")\n",
    "        \n",
    "        # add the weighted effect size to the prs\n",
    "        prs_score += effect_size * genotype\n",
    "    \n",
    "    return prs_score\n",
    "\n",
    "# calculate the prs with the correct column names\n",
    "prs = calculate_prs(gwas_summary, genotype_data_dict, p_value_col, snp_col=snp_col, effect_size_col=effect_size_col)\n",
    "print(f\"Polygenic Risk Score: {prs}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3394c6a4-d8d5-419d-a2f0-bdbe2af6db66",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
